authentication/oidc.go

```diff
@@ -44,54 +44,45 @@ type OIDCConfig struct {
 // Middleware is a convenience type for functions that wrap http.Handlers.
 type Middleware func(http.Handler) http.Handler
 
-// NewOIDC creates a single http.Handler and a set of Middlewares for all
-// tenants that is able to authenticate requests and provide the
+// NewOIDC creates a http.Handler and a Middlewares for a
+// tenant that is able to authenticate requests and provide the
 // authorization code grant flow for users.
-func NewOIDC(logger log.Logger, prefix string, configs []TenantOIDCConfig) (http.Handler, map[string]Middleware, []error) {
-	handlers := map[string]http.Handler{}
-	middlewares := map[string]Middleware{}
-	warnings := make([]error, 0, len(configs))
+func NewOIDC(logger log.Logger, prefix string, config TenantOIDCConfig) (http.Handler, Middleware, error) {
+	p, err := NewProvider(context.TODO(), logger, getCookieForTenant(config.Tenant), "/"+config.Tenant, config.OIDCConfig)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	for _, c := range configs {
-		p, err := NewProvider(context.TODO(), logger, getCookieForTenant(c.Tenant), "/"+c.Tenant, c.OIDCConfig)
-		if err != nil {
-			warnings = append(warnings, fmt.Errorf("failed to instantiate OIDC provider for tenant %q: %w", c.Tenant, err))
-			continue
-		}
+	h := chi.NewRouter()
 
-		r := chi.NewRouter()
+	const (
+		loginRoute    = "/login"
+		callbackRoute = "/callback"
+	)
 
-		const (
-			loginRoute    = "/login"
-			callbackRoute = "/callback"
-		)
+	h.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
+	h.Handle(callbackRoute, otelhttp.WithRouteTag(prefix+callbackRoute, p.CallbackHandler()))
 
-		r.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
-		r.Handle(callbackRoute, otelhttp.WithRouteTag(prefix+callbackRoute, p.CallbackHandler()))
-
-		handlers[c.Tenant] = r
-		middlewares[c.Tenant] = p.Middleware()
-	}
+	middleware := p.Middleware()
 
 	r := chi.NewRouter()
 	r.Mount("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tenant, ok := GetTenant(r.Context())
-		const msg = "error finding tenant"
+		_, ok := GetTenant(r.Context())
+		const msg = "P error finding tenant"
 		if !ok {
 			level.Warn(logger).Log("msg", msg)
 			http.Error(w, msg, http.StatusInternalServerError)
 			return
 		}
-		h, ok := handlers[tenant]
-		if !ok {
+		if r == nil {
 			level.Debug(logger).Log("msg", msg)
 			http.Error(w, msg, http.StatusUnauthorized)
 			return
 		}
 		h.ServeHTTP(w, r)
 	}))
 
-	return r, middlewares, warnings
+	return h, middleware, nil
```

Comment:
The router named `r` seems completely unused. We create it, mount a handler and then never return it. This router was composing the handlers for all of the tenants. Maybe it's not needed anymore given your changes?

Reply:
router 'r' is used to direct the traffic to the right tenant (one level of indirection). Yes, 'r' is not used, removed the code to create this extra layer of redirection which is not required if we can deal with handler for the tenant that the traffic is destined to.